cpp/inc/bond/core/detail/any.h

```diff
@@ -0,0 +1,307 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+/*
+ * This header implements bond::detail::any class which provides similar functionality to
+ * the boost::any which (1) uses user-defined CTTI instead of a built-in RTTI and (2) uses
+ * user-defined storage size for small object optimization.
+ */
+
+
+#pragma once
+
+#include "mpl.h"
+#include <bond/core/config.h>
+#include <boost/assert.hpp>
+#include <boost/static_assert.hpp>
+#include <boost/utility/enable_if.hpp>
+#include <cstdint>
+#include <type_traits>
+
+
+namespace bond
+{
+namespace detail
+{
+
+template <template <typename> class TypeId, std::size_t Size>
+class any
+{
+    BOOST_STATIC_ASSERT(Size >= sizeof(void*));
+
+public:
+    any() BOND_NOEXCEPT
+        : _id{}
+    {}
+
+    any(const any& other)
+    {
+        emplace(other);
+    }
+
+    template <typename T>
+    any(const T& value)
+    {
+        emplace(value);
+    }
+
+    ~any()
+    {
+        destroy();
+    }
+
+    any& operator=(const any& other)
+    {
+        return assign(other);
+    }
+
+    template <typename T>
+    any& operator=(const T& value)
+    {
+        return assign(value);
+    }
+
+    bool empty() const BOND_NOEXCEPT
+    {
+        return _id == 0;
+    }
+
+    explicit operator bool() const BOND_NOEXCEPT
+    {
+        return !empty();
+    }
+
+    template <typename T>
+    T* cast() BOND_NOEXCEPT
+    {
+        return TypeId<T>::value == _id ? functions::template table<T>::unsafe_cast(*this) : nullptr;
+    }
+
+    template <typename T>
+    const T* cast() const BOND_NOEXCEPT
+    {
+        return TypeId<T>::value == _id ? functions::template table<T>::unsafe_cast(*this) : nullptr;
+    }
+
+    bool operator==(const any& other) const
+    {
+        return _id == other._id && (empty() || _functions.compare(*this, other));
+    }
+
+    bool operator!=(const any& other) const
+    {
+        return !(*this == other);
+    }
+
+private:
+    using storage = typename std::aligned_storage<Size>::type;
+
+    struct functions
+    {
+        void (*destroy)(any& x);
+
+        void (*assign)(any& x, const any& other);
+
+        void (*emplace)(any& x, const any& other);
+
+        bool (*compare)(const any& x, const any& y);
+
+
+        template <typename T, bool IsSmall>
+        struct impl;
+
+        template <typename T>
+        struct impl<T, true>
+        {
+            static T* unsafe_cast(any& x) BOND_NOEXCEPT
+            {
+                BOOST_ASSERT(TypeId<T>::value == x._id);
+                return reinterpret_cast<T*>(x.data());
```

Comment:
Any reason we can't use `static_cast<T*>` here and below? `x.data()` returns a `void*`.